[ONNX] Export new_empty_tensor

neginraoof opened the pull request:
Symbolic for exporting new_empty_tenor to onnx

Currently, ORT does not support shape value 0.

lara-hdr commented:
@fmassa, 
@neginraoof added this PR to deal with cases where ConvTranspose2d's input is empty, and the registered op _new_empty_tensor is called (https://github.com/pytorch/vision/blob/d88d8961ae51507d0cb680329d985b1488b1b76b/torchvision/ops/misc.py#L46).

We are seeing a case in maskrcnn where _new_empty_tensor is called and creates an output with a 0 for one of its dimensions.

@fmassa, could you give a bit of context on when this could happen? Is it when nothing is found in the image?

@neginraoof please add more info if I missed something :)

neginraoof commented:
Thanks @lara-hdr for the comments.

For Mask RCNN and Keypoint RCNN, when we test the models with random input, and therefore there is no detection, this part of the code in ConvTranspose2d is hit: https://github.com/pytorch/vision/blob/61763fa955ef74077a1d3e1aa5da36f7c606943a/torchvision/ops/misc.py#L45
Where the input is empty, and hence output shape has a dim value 0.

fmassa commented:
Hi all,

This could happen when no detections have been found in the image. This is not an issue for Faster R-CNN, but could happen for Mask R-CNN and Keypoint R-CNN.

In PyTorch, with https://github.com/pytorch/pytorch/pull/31658 we will have proper support for 0-sized batches in PyTorch, so we will be getting closer to removing `_new_empty_tensor`, but for now this is needed.

fmassa commented on test/test_onnx.py:
```diff
@@ -74,6 +74,18 @@ def to_numpy(tensor):
+            def forward(self, input2):
+                return self.conv2(input2)
+
+        input = torch.empty(0)
```
Can you make `input` be instead a 4d tensor, with proper shapes?
Like
```python
input = torch.rand(0, 16, 10, 10)
```
this way we make sure that the tests work as expected. Indeed, transposed convolution only support tensors with a batch size of zero, the other dimensions should be valid

fmassa commented on test/test_onnx.py:
```diff
@@ -74,6 +74,18 @@ def to_numpy(tensor):
+                return self.conv2(input2)
+
+        input = torch.empty(0)
+        self.run_model(Module(), [(input, )])
```
Can you run the model with two different inputs with different shapes, to verify that dynamic shapes work?

neginraoof commented:
Thanks @fmassa and @lara-hdr 
I addressed the feedback.
This PR is pending a couple other PR's in PT and ONNX/ORT:
A small fix in PT op registration API: https://github.com/pytorch/pytorch/pull/32017
Fix for onnx::split with zero lenghts: https://github.com/onnx/onnx/pull/2544
After the ONNX PR is merged, I'll look into the changes required for split in ORT.

fmassa commented:
@neginraoof thanks for the update!

Let us know when the pending PRs are merged, so that we can merge this

neginraoof commented:
@fmassa This PR is ready for review now. I will enable the ORT test once ORT implements split for empty tensor.
Thanks.
